### PR TITLE
[YUNIKORN-2494] Revisit IsAtorAbove, WithIn, GetRemaining Guaranteed …

### PR DESCRIPTION
**Pull request by manirajv06**
…resources calculation

### What is this PR for?
Methods like  IsAtorAbove, WithIn etc doesn't expose the actual guaranteed values and returns boolean value based on the calculation. There are cases, where these boolean values are not correct and also there is a need to know the actual guaranteed values. For example, How much is remaining in Guaranteed? How much can be preempted? etc.

Introduced new methods to expose the actual values and incorporated the same into existing preemption flow. Added tests cases to cover the new methods as well.

### What type of PR is it?
* [ ] - Improvement

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2494

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by manirajv06** [comment]
Have decided to split the pr into two separate tasks. Raised https://github.com/apache/yunikorn-core/pull/822 containing new methods and its test cases. Filed https://issues.apache.org/jira/browse/YUNIKORN-2500 to track incorporating these new methods into existing preemption code.

File: pkg/scheduler/objects/preemption.go [hunk 1]
```diff
@@ -692,77 +696,58 @@ func (qps *QueuePreemptionSnapshot) Duplicate(copy map[string]*QueuePreemptionSn
...
-func (qps *QueuePreemptionSnapshot) IsWithinGuaranteedResource() bool {
-	if qps == nil {
-		return true
+	used.SubOnlyExisting(qps.PreemptingResource)
+
+	// Calculate preemptable resource. +ve means Over utilized, -ve means Under utilized, 0 means correct utilization
+	guaranteed := qps.GuaranteedResource.Clone()
+	actualPreemptableResource := used.Clone()
+	actualPreemptableResource.SubOnlyExisting(guaranteed)
+	preemptableResource := actualPreemptableResource.Clone()
```
Comment (pbacsko) [inline pkg/scheduler/objects/preemption.go:716]: Is copy necessary? You can delete keys during iteration.

File: pkg/scheduler/objects/preemption_test.go [hunk 2]
```diff
@@ -166,6 +166,7 @@ func TestCheckPreemptionQueueGuaranteesWithNoGuaranteedResources(t *testing.T) {
 			ask3 := newAllocationAsk("alloc3", appID2, resources.NewResourceFromMap(map[string]resources.Quantity{"first": 5}))
 			assert.NilError(t, app2.AddAllocationAsk(ask3))
 			childQ2.incPendingResource(ask3.GetAllocatedResource())
+			print("ff is " + childQ2.GetAllocatedResource().String())
```
Comment (pbacsko) [inline pkg/scheduler/objects/preemption_test.go:169]: Is this needed?

**Comment by pbacsko** [comment]
ping @manirajv06 what's the matter with this one? There was another PR #822 with the same name.

**Comment by manirajv06** [comment]
> ping @manirajv06 what's the matter with this one? There was another PR #822 with the same name.

@pbacsko Yes, have raised another PR #822. As explained above, split this PR into two different pr's. Will close this one.

**Pull request closed without merging**